spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/GcpSecretManagerBootstrapConfiguration.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.secretmanager;
+
+import java.io.IOException;
+
+import javax.annotation.PostConstruct;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceSettings;
+import com.google.protobuf.ByteString;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.convert.ApplicationConversionService;
+import org.springframework.cloud.bootstrap.config.PropertySourceLocator;
+import org.springframework.cloud.gcp.core.DefaultCredentialsProvider;
+import org.springframework.cloud.gcp.core.DefaultGcpProjectIdProvider;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.cloud.gcp.core.UserAgentHeaderProvider;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.core.convert.converter.ConverterRegistry;
+
+/**
+ * Bootstrap Autoconfiguration for GCP Secret Manager which enables loading secrets as
+ * properties into the application {@link org.springframework.core.env.Environment}.
+ *
+ * @author Daniel Zou
+ * @since 1.3
+ */
+@Configuration
+@EnableConfigurationProperties(GcpSecretManagerProperties.class)
+@ConditionalOnClass(SecretManagerServiceClient.class)
+@ConditionalOnProperty(value = "spring.cloud.gcp.secretmanager.enabled", matchIfMissing = true)
+public class GcpSecretManagerBootstrapConfiguration {
+
+	private final GcpSecretManagerProperties properties;
+
+	private final CredentialsProvider credentialsProvider;
+
+	private final GcpProjectIdProvider gcpProjectIdProvider;
+
+	public GcpSecretManagerBootstrapConfiguration(GcpSecretManagerProperties properties)
+			throws IOException {
+
+		this.properties = properties;
+		this.credentialsProvider = new DefaultCredentialsProvider(properties);
+		this.gcpProjectIdProvider = properties.getProjectId() != null
+				? properties::getProjectId
+				: new DefaultGcpProjectIdProvider();
+	}
+
+	@PostConstruct
+	public void init() {
+		// In this method, we register the converters which convert SecretPayload objects to
+		// Strings and byte[].
+		ConverterRegistry converterRegistry = (ConverterRegistry) ApplicationConversionService.getSharedInstance();
```

Comment:
Asking the boot team if there is a better way than using a static method.

Reply:
If I'm not mistaken, you should just be able to register these as beans.

Reply:
Unfortunately this does not work; I did something like this:

```
public class GcpSecretManagerBootstrapConfiguration {

  ... other beans ...

	@Bean
	public Converter<ByteString, String> getConverterA() {
		return new Converter<ByteString, String>() {
			@Override
			public String convert(ByteString source) {
				return source.toStringUtf8();
			}
		};
	}

	@Bean
	public Converter<ByteString, byte[]> getConverterB() {
		return new Converter<ByteString, byte[]>() {
			@Override
			public byte[] convert(ByteString source) {
				return source.toByteArray();
			}
		};
	}
}
```

Reply:
Hmm, maybe injecting `ConfigurableEnvironment` and using `ConfigurableEnvironment.getConversionService()` rather than the static method then?

Reply:
I think the beans have to be annotated with `@ConfigurationPropertiesBinding` to be discoverable.

Reply:
that's specific to binding and `ByteString` isn't used in binding.

Reply:
The `ConfigurableEnvironment` method worked for me!